### PR TITLE
Fix freeze and traverse utility methods

### PR DESCRIPTION
**Pull request by awav**
TensorFlow Probability 0.10 introduces a new `_parameters` attribute in tfp.bijectors.Bijector instances that contains a self-reference, which breaks GPflow's module-traversion utilities (leaf_components, print_summary, deepcopy, freeze). This PR changes `traverse_module` to ignore the `_parameters` attribute and proposes a new deepcopy approach that simply uses copy.deepcopy()'s `memo` argument.

@st-- has this solution #1473. This PR provides a less hacky and improved approach.

### CODE REVIEW

File: doc/source/notebooks/intro_to_gpflow2.pct.py [hunk 1]
```diff
@@ -564,3 +563,7 @@ def checkpointing_training_loop(
 
 p = gpflow.Parameter(1.1, transform=gpflow.utilities.positive())
 print(f"{global_str}{p}")
+
+# %%
+
+# %%
```
Comment (st--) [inline doc/source/notebooks/intro_to_gpflow2.pct.py:569]:
```suggestion
```

File: gpflow/utilities/utilities.py [hunk 2]
```diff
@@ -382,12 +383,9 @@ def freeze(input_module: M) -> M:
...
     """
-    module_copy = deepcopy(input_module)
-    for name, value in parameter_dict(module_copy).items():
-        tensor = tf.convert_to_tensor(value, dtype=value.dtype)
-        const_value = tf.constant(tensor)
-        attr_path = name.lstrip(".")
-        setattr_by_path(module_copy, attr_path, const_value)
+    objects_to_freeze = _get_leaf_components(input_module)
+    memo_tensors = {id(v): tf.constant(tf.convert_to_tensor(v)) for v in objects_to_freeze.values()}
+    module_copy = deepcopy(input_module, memo_tensors)
```
[the remark this answers is not in the document]
Comment (awav) [reply gpflow/utilities/utilities.py:388]: Yes, existing tests fit well. This method is how we were supposed to do `deepcopy` in the first place.

File: doc/source/notebooks/intro_to_gpflow2.pct.py [hunk 3]
```diff
@@ -510,17 +510,16 @@ def checkpointing_training_loop(
...
+    frozen_model.predict_f, input_signature=[tf.TensorSpec(shape=[None, 1], dtype=tf.float64)]
 )
 module_to_save.predict = predict_fn
 
 # %% [markdown]
 # Save original result for futher comparison
 
 # %%
-original_result = module_to_save.predict(samples_input)
+sample_input_tensor = tf.convert_to_tensor(samples_input, dtype=default_float())
```
Comment (st--) [inline doc/source/notebooks/intro_to_gpflow2.pct.py:521]: Could you add a comment explaining why you convert to tensor first ?

File: gpflow/utilities/utilities.py [hunk 4]
```diff
@@ -421,7 +419,13 @@ def traverse_module(
             new_state = traverse_module(subterm, new_acc, update_cb, target_types)
     elif isinstance(m, tf.Module):
         for name, submodule in vars(m).items():
-            if name in tf.Module._TF_MODULE_IGNORED_PROPERTIES:
+            ignore = m._TF_MODULE_IGNORED_PROPERTIES
```
Comment (st--) [inline gpflow/utilities/utilities.py:422]:
```suggestion
            ignored_attributes = m._TF_MODULE_IGNORED_PROPERTIES
```

File: gpflow/utilities/utilities.py [hunk 5]
```diff
@@ -421,7 +419,13 @@ def traverse_module(
...
+            # NOTE(awav): since tfp version 0.10.0, tfp.bijectors.Bijector instances have
+            # `_parameters` dictionary with `self` references that cause
+            # infinite recursive loop.
+            if isinstance(m, tfp.bijectors.Bijector):
+                ignore = ignore.union(["_parameters"])
```
Comment (st--) [inline gpflow/utilities/utilities.py:427]:
```suggestion
                ignored_attributes = ignored_attributes.union({"_parameters"})
```

File: gpflow/utilities/utilities.py [hunk 6]
```diff
@@ -421,7 +419,13 @@ def traverse_module(
...
+            if name in ignore:
```
Comment (st--) [inline gpflow/utilities/utilities.py:428]:
```suggestion
            if name in ignored_attributes:
```

File: gpflow/utilities/utilities.py [hunk 7]
```diff
@@ -421,7 +419,13 @@ def traverse_module(
             new_state = traverse_module(subterm, new_acc, update_cb, target_types)
     elif isinstance(m, tf.Module):
         for name, submodule in vars(m).items():
-            if name in tf.Module._TF_MODULE_IGNORED_PROPERTIES:
+            ignore = m._TF_MODULE_IGNORED_PROPERTIES
+            # NOTE(awav): since tfp version 0.10.0, tfp.bijectors.Bijector instances have
+            # `_parameters` dictionary with `self` references that cause
```
Comment (st--) [inline gpflow/utilities/utilities.py:424]:
```suggestion
            # `_parameters` dictionary with "self" references that cause
```

File: gpflow/utilities/utilities.py [hunk 8]
```diff
@@ -382,12 +383,9 @@ def freeze(input_module: M) -> M:
...
     :return: Returns a frozen deepcopy of an input object.
     """
-    module_copy = deepcopy(input_module)
-    for name, value in parameter_dict(module_copy).items():
-        tensor = tf.convert_to_tensor(value, dtype=value.dtype)
-        const_value = tf.constant(tensor)
-        attr_path = name.lstrip(".")
-        setattr_by_path(module_copy, attr_path, const_value)
+    objects_to_freeze = _get_leaf_components(input_module)
+    memo_tensors = {id(v): tf.constant(tf.convert_to_tensor(v)) for v in objects_to_freeze.values()}
```
Comment (st--) [inline gpflow/utilities/utilities.py:387]:
```suggestion
    memo_tensors = {id(v): tf.convert_to_tensor(v) for v in objects_to_freeze.values()}
```
the tf.constant() doesn't do anything:
```python
a = tf.convert_to_tensor(1.0)
b = tf.constant(a)
assert a is b
```

File: gpflow/utilities/utilities.py [hunk 9]
```diff
@@ -364,15 +364,16 @@ def setattr_by_path(target: object, attr_path: str, value: Any):
...
 
 
-def deepcopy(input_module: M) -> M:
+def deepcopy(input_module: M, memo: Optional[Dict[int, Any]] = None) -> M:
     """
     Returns a deepcopy of the input tf.Module. To do that first resets the caches stored inside each
     tfp.bijectors.Bijector to allow the deepcopy of the tf.Module.
 
     :param input_module: tf.Module including keras.Model, keras.layers.Layer and gpflow.Module.
+    :param memo: Check details for func:`copy.deepcopy` `memo` argument.
```
Comment (st--) [inline gpflow/utilities/utilities.py:373]:
```suggestion
    :param memo: passed through to func:`copy.deepcopy` (see https://docs.python.org/3/library/copy.html).
```
The docstring of the `copy` module and `copy.deepcopy` itself are actually rather unhelpful...

**Comment by st--** [comment]
Can we delete the whole `set_attr_by_path` set of functions now ?

**Comment by awav** [comment]
> Can we delete the whole `set_attr_by_path` set of functions now ?

Yes, I did.

**Comment by st--** [comment]
@awav do we also want to get rid of the helper functions such as `_get_by_name_index`?

File: gpflow/utilities/utilities.py [hunk 10]
```diff
@@ -382,12 +343,9 @@ def freeze(input_module: M) -> M:
     :param input_module: tf.Module or gpflow.Module.
     :return: Returns a frozen deepcopy of an input object.
     """
-    module_copy = deepcopy(input_module)
-    for name, value in parameter_dict(module_copy).items():
-        tensor = tf.convert_to_tensor(value, dtype=value.dtype)
-        const_value = tf.constant(tensor)
-        attr_path = name.lstrip(".")
-        setattr_by_path(module_copy, attr_path, const_value)
+    objects_to_freeze = _get_leaf_components(input_module)
```
Comment (st--) [inline gpflow/utilities/utilities.py:346]:
```suggestion
    objects_to_freeze = leaf_components(input_module)
```
why are there two ways of calling it, anyways?

**Comment by awav** [comment]
> @awav do we also want to get rid of the helper functions such as `_get_by_name_index`?

There is no much use in this method. It is too generic for gpflow after upgrading the deepcopy.

File: gpflow/utilities/utilities.py (hunk 10 above)
Comment (awav) [reply gpflow/utilities/utilities.py:346]: I prefer to leave this call to a private module method. This method was never supposed to be exposed. For some reason it was. Let's discuss it outside of this PR.

**Comment by st--** [comment]
> There is no much use in this method. It is too generic for gpflow after upgrading the deepcopy.

Exactly, but they're still in the code, that's why I asked - to clarify, as you already removed `[gs]et_attr_by_path`, please also remove the now-no-longer-used helper functions :)

File: doc/source/notebooks/intro_to_gpflow2.pct.py [hunk 11]
```diff
@@ -510,17 +510,16 @@ def checkpointing_training_loop(
...
+original_result = module_to_save.predict(sample_input_tensor)
```
Comment (awav) [inline doc/source/notebooks/intro_to_gpflow2.pct.py:522]:
```suggestion
original_result = module_to_save.predict(samples_input)
```

**Comment by awav** [comment]
> > There is no much use in this method. It is too generic for gpflow after upgrading the deepcopy.
> 
> Exactly, but they're still in the code, that's why I asked - to clarify, as you already removed `[gs]et_attr_by_path`, please also remove the now-no-longer-used helper functions :)

if you have a list, please share it with me.

**Comment by awav** [comment]
> > There is no much use in this method. It is too generic for gpflow after upgrading the deepcopy.
> 
> Exactly, but they're still in the code, that's why I asked - to clarify, as you already removed `[gs]et_attr_by_path`, please also remove the now-no-longer-used helper functions :)

Apologies, I was confused by naming. It is done.

File: doc/source/notebooks/intro_to_gpflow2.pct.py (hunk 3 above)
Comment (awav) [reply doc/source/notebooks/intro_to_gpflow2.pct.py]: I changed it during debugging and found out that using `np.array` with `tf.function` here is a bad example and I left it.

**Pull request merged**